website/blog/2020-03-02-presto-lambda.md

```diff
@@ -0,0 +1,194 @@
+
+The type parameter `T` and `U` needs to be determined to resolve the expression type. This is done by [SignaturerBinder#bind](https://github.com/prestodb/presto/blob/5833338c127c380f505873614d25862921437e75/presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java#L92) method. [SignaturerBinder#appendConstraintSolvers](https://github.com/prestodb/presto/blob/5833338c127c380f505873614d25862921437e75/presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java#L289-L300) is called under the hood to iteratively solve this constraint satisfaction problem. A new [TypeConstraintSolver](https://github.com/prestodb/presto/blob/5833338c127c380f505873614d25862921437e75/presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java#L541-L544) called [FunctionSolver](https://github.com/prestodb/presto/blob/5833338c127c380f505873614d25862921437e75/presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java#L720-L721) was added for updating type constraints related to lambda expressions.
+
+## Lambda Capture
+
+Lambda capture allows users to refer to other columns in the lambda function, for example: 
+
+```sql
+SELECT elements, 
+    transform(elements, v -> v * factor) as multiplied_elements
+FROM (
+    VALUES
+        (ARRAY[1, 2], 2),
+        (ARRAY[1, 3, 9], 3),
+        (ARRAY[1, 4, 16, 64], 4)
+) AS t(elements, factor)
+
+    elements    | multiplied_elements
+----------------+---------------------
+ [1, 2]         | [2, 4]
+ [1, 3, 9]      | [3, 9, 27]
+ [1, 4, 16, 64] | [4, 16, 64, 256]
+(3 rows)
+```
+
+Lambda capture supported is added in [Presto#7210](https://github.com/prestodb/presto/pull/7210). It rewrites the captured lambda into non-capture lambda via [partial function application](https://en.wikipedia.org/wiki/Partial_application). A special internal higher order function `BIND` is introduced to partially apply captured arguments to the lambda.
+
+Take the above example, the captured lambda call
+
+```sql
+transform(elements, v -> v * factor)
+```
+
+is rewritten to
+
+```sql
+transform(
+    elements, 
+    BIND(factor, (captured_factor, e) -> e * captured_factor)
+)
+```
+
+The original unary lambda with capture `e -> e * factor` is rewritten into a binary lambda without capture: `(captured_factor, e) -> e * captured_factor`. The `BIND` call takes `factor` and this binary lambda as input, returns the partially applied function that multiplies the input by `captured_factor` (Note the `captured_factor` will be different for each row!). This partially applied function is a unary function and is provided as the second parameter to `transform` call.
+
+## Lambda Execution
+
+In this section we are going to discuss how lambda is executed during runtime. The original implementation used `MethodHandle` object to represent lambda on stack. Consider the same example:
```

Comment:
how lambda -> how a lambda
"used `MethodHandle` object to represent lambda on stack" -> "used `MethodHandle` objects to represent lambdas on the stack"